src/main/java/com/google/devtools/build/lib/remote/http/HttpDownloadHandler.java

```diff
@@ -159,16 +186,36 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable t) {
     }
   }
 
-  private HttpRequest buildRequest(String path, String host) {
+  private HttpRequest buildRequest(String path, String host, long offset) {
     HttpRequest httpRequest =
         new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, path);
     httpRequest.headers().set(HttpHeaderNames.HOST, host);
     httpRequest.headers().set(HttpHeaderNames.CONNECTION, HttpHeaderValues.KEEP_ALIVE);
     httpRequest.headers().set(HttpHeaderNames.ACCEPT, "*/*");
     httpRequest.headers().set(HttpHeaderNames.ACCEPT_ENCODING, HttpHeaderValues.GZIP);
+    if (offset != 0) {
+      httpRequest.headers().set(HttpHeaderNames.RANGE, String.format("%s=%d-", HttpHeaderValues.BYTES, offset));
+    }
```

Comment:
How is the RANGE implementation in this PR working in combination with the also supplied `ACCEPT-ENCODING: gzip` header?

It seems this PR expects ranges to be applied to the uncompressed data, right? but RFC 9110 states: _“If the representation data has a content coding applied, each byte range is calculated with respect to the encoded sequence of bytes, not the sequence of underlying bytes that would be obtained after decoding.“_

For the gPRC cache protocol, there were lots of discussions and considerations in https://github.com/bazelbuild/remote-apis/pull/168 about combining compression with read offsets. E.g. regarding if offsets should be applied to the compressed or non-compressed stream, regarding when offsets don’t align with stored compressed chunk sizes on the server side, etc.

How often do you expect retries of already partially downloaded files? **Would it be sufficient to always re-request the whole file and simply skip already received CAS chunks after receiving them again?** Would that avoid much of the complexities in both code and test cases?

[Bazel-remote](https://github.com/buchgr/bazel-remote) supports zstd compression for both the gRPC and HTTP protocols. Bazel is still only supporting it for gRPC. I’m afraid the RANGE implementation part of this PR would make it harder for future work of adding support also for `ACCEPT-ENCODING: zstd` to the HTTP cache client.

Reply:
Correct, the offset is calculated on uncompressed bytes.

> Bazel is still only supporting it for gRPC. I’m afraid the RANGE implementation part of this PR would make it harder for future work of adding support also for ACCEPT-ENCODING: zstd to the HTTP cache client.

I see, fair enough. Perhaps a solution that takes encoding into account could introduce a new step [into the download pipeline before the inflater](https://github.com/bazelbuild/bazel/blob/5e763b4bd25a7a9991953b79beeaef4852f77c5e/src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java#L399) to determine the correct download offset in light of encoding.

> How often do you expect retries of already partially downloaded files? **Would it be sufficient to always re-request the whole file and simply skip already received CAS chunks after receiving them again?** Would that avoid much of the complexities in both code and test cases?

The failures were sporadic, usually a handful in one CI pipeline. So, this may well be good enough. That would help make the tests simpler as the mock server wouldn't have to worry about range requests. I'll try this route.